xapian-core/include/xapian/cluster.h

```diff
@@ -0,0 +1,603 @@
+     */
+    PointType * release() {
+	opt_intrusive_base::release();
+	return this;
+    }
+
+    /** Start reference counting this object.
+     *
+     *  You can hand ownership of a dynamically allocated PointType
+     *  object to Xapian by calling release() and then passing the object to a
+     *  Xapian method.  Xapian will arrange to delete the object once it is no
+     *  longer required.
+     */
+    const PointType * release() const {
+	opt_intrusive_base::release();
+	return this;
+    }
+};
+
+/** Class to represent a document as a point in the Vector Space
+ *  Model
+ */
+class XAPIAN_VISIBILITY_DEFAULT Point : public PointType {
+    /// The document which is being represented by the Point
+    Document document;
+
+  public:
+    /** Constructor
+     *  Initialize the point with terms and corresponding TF-IDF weights
+     *
+     *  @param tlg		TermListGroup object which provides the term
+     *				frequencies which is used for TF-IDF weight calulations
+     *  @param document		The Document object over which the Point object
+     *				will be initialized
+     */
+    Point(const TermListGroup &tlg, const Document &document);
+
+    /// Returns the document corresponding to this Point
+    Document get_document() const;
+};
+
+/** Class to represent cluster centroids in the vector space
+*/
+class XAPIAN_VISIBILITY_DEFAULT Centroid : public PointType {
+  public:
+    // Constructor
+    Centroid();
+
+    /** Constructor with Point argument
+     *
+     *  @param point	Point object to which Centroid object is
+     *			initialized. The document vector and the
+     *			magnitude are made equal
+     */
+    explicit Centroid(const Point &point);
+
+    /** Divide the weight of terms in the centroid by 'size' and
+     *  recalculate the magnitude
+     *
+     *  @param cluster_size	Value by which Centroid document vector is
+     *				divided
+     */
+    void divide(double cluster_size);
+
+    /// Clear the terms and corresponding values of the centroid
+    void clear();
+
+    /// Recalculate the magnitude of the centroid
+    void recalc_magnitude();
+};
+
+/** Class to represents a Cluster which contains Points and
+ *  of the Cluster
+ */
+class XAPIAN_VISIBILITY_DEFAULT Cluster {
+  public:
+    class Internal;
+    /// @private @internal Reference counted internals.
+    Xapian::Internal::intrusive_ptr<Internal> internal;
+
+    /** Copying is allowed.  The internals are reference counted, so
+     *  copying is cheap.
+     *
+     *  @param other	The object to copy.
+     */
+    Cluster(const Cluster &other);
+
+    /** Assignment is allowed.  The internals are reference counted,
+     *  so assignment is cheap.
+     *
+     *  @param other	The object to copy.
+     */
+    Cluster& operator=(const Cluster &other);
+
+    /** Constructor
+     *
+     *  @param centroid		The centroid of the cluster object is
+     *				assigned to 'centroid'
+     */
+    Cluster(const Centroid &centroid);
```

Comment:
This constructor should be marked as `explicit` to prevent implicit conversion of `Centroid` to `Cluster`.